issue with pod restart while waiting for condition

if we are waiting for a condition, e.g. 'check that container X has Z in its logs' and the pod for that container gets reset during the wait block, the test could fail with an error similar to:

```
E           kubernetes.client.rest.ApiException: (400)
E           Reason: Bad Request
E           HTTP response headers: HTTPHeaderDict({'Audit-Id': '3f90ef9f-b1b6-46e8-97e7-87854e36801e', 'Content-Type': 'application/json', 'Date': 'Mon, 24 Sep 2018 19:07:59 GMT', 'Content-Length': '213'})
E           HTTP response body: {"kind":"Status","apiVersion":"v1","metadata":{},"status":"Failure","message":"container \"blackbox\" in pod \"blackbox-666cffd4c9-8lt4b\" is waiting to start: ContainerCreating","reason":"BadRequest","code":400}
```

this happens because while we are re-checking for the condition, the pod restarted and is temporarily unavailable.

I think the solution to this is to add a param to the `wait` function that, when set, will prevent these failures from erroring out of the check loop.

--- a/kubetest/client.py
+++ b/kubetest/client.py
@@ -357,7 +357,9 @@ class TestClient:
     # ****** Test Helpers ******
 
     @staticmethod
-    def wait_for_conditions(*args, timeout=None, interval=1, policy=Policy.ONCE):
+    def wait_for_conditions(
+            *args, timeout=None, interval=1, policy=Policy.ONCE, fail_on_api_error=True
+    ):
         """Wait for all of the provided Conditions to be met.
 
         All Conditions must be met for this to unblock. If no Conditions are
@@ -373,6 +375,12 @@ class TestClient:
                 re-evaluating the conditions. Default: 1s
             policy (condition.Policy): The condition checking policy that defines
                 the checking behavior. Default: ONCE
+            fail_on_api_error (bool): Fail the condition checks if a Kubernetes
+                API error is incurred. An API error can be raised for a number
+                of reasons, including a Pod being restarted and temporarily
+                unavailable. Disabling this will cause those errors to be
+                ignored, allowing the check to continue until timeout or
+                resolution. (default: True).
 
         Raises:
             TimeoutError: The Conditions were not met within the specified
@@ -418,6 +426,7 @@ class TestClient:
                 condition=wait_condition,
                 timeout=timeout,
                 interval=interval,
+                fail_on_api_error=fail_on_api_error,
             )
         except TimeoutError:
             # If we time out here, we want to show all the conditions

--- a/kubetest/objects/pod.py
+++ b/kubetest/objects/pod.py
@@ -1,5 +1,6 @@
 """Kubetest wrapper for the Kubernetes `Pod` API Object."""
 
+import json
 import logging
 
 from kubernetes import client
@@ -167,6 +168,35 @@ class Pod(ApiObject):
             total += container_status.restart_count
 
         return total
+
+    def http_proxy_get(self, path):
+        """Issue a GET request to a proxy for the Pod.
+
+        Args:
+            path (str): The URI to hit on the Pod.
+
+        Returns:
+            str: The response data.
+        """
+        return client.CoreV1Api().connect_get_namespaced_pod_proxy_with_path(
+            name=self.name,
+            namespace=self.namespace,
+            path=path,
+        )
+
+    def http_proxy_get_json(self, path):
+        """Issue a GET request to a proxy for the Pod and return
+        the JSON response as a dictionary.
+
+        Args:
+            path (str): The URI to hit on the Pod.
+
+        Returns:
+            dict: The response data.
+        """
+        data = self.http_proxy_get(path)
+        data = data.replace("'", '"')
+        return json.loads(data)
 
     def containers_started(self):
         """Check if the Pod's Containers have all started.

--- a/kubetest/utils.py
+++ b/kubetest/utils.py
@@ -3,6 +3,8 @@
 import logging
 import time
 
+from kubernetes.client.rest import ApiException
+
 log = logging.getLogger('kubetest')
 
 
@@ -64,7 +66,7 @@ def selector_kwargs(fields=None, labels=None):
     return kwargs
 
 
-def wait_for_condition(condition, timeout=None, interval=1):
+def wait_for_condition(condition, timeout=None, interval=1, fail_on_api_error=True):
     """Wait for a condition to be met.
 
     Args:
@@ -75,6 +77,12 @@ def wait_for_condition(condition, timeout=None, interval=1):
             or exceeded, a TimeoutError will be raised.
         interval (int|float): The time, in seconds, to wait before
             re-checking the condition.
+        fail_on_api_error (bool): Fail the condition checks if a Kubernetes
+            API error is incurred. An API error can be raised for a number
+            of reasons, including a Pod being restarted and temporarily
+            unavailable. Disabling this will cause those errors to be
+            ignored, allowing the check to continue until timeout or
+            resolution. (default: True).
 
     Raises:
         TimeoutError: The specified timeout was exceeded.
@@ -97,8 +105,13 @@ def wait_for_condition(condition, timeout=None, interval=1):
             )
 
         # check if the condition is met and break out if it is
-        if condition.check():
-            break
+        try:
+            if condition.check():
+                break
+        except ApiException as e:
+            log.warning('got api exception while waiting: {}'.format(e))
+            if fail_on_api_error:
+                raise
 
         # if the condition is not met, sleep for the interval
         # to re-check later

